README.md
# baucis

[![lerna](https://img.shields.io/badge/maintained%20with-lerna-cc00ff.svg)](https://lernajs.io/)
[![Npm version](https://img.shields.io/npm/v/@coorpacademy/baucis.svg)](https://www.npmjs.com/package/@coorpacademy/baucis)
[![Build Status](https://travis-ci.com/CoorpAcademy/baucis.svg?branch=master)](https://travis-ci.com/CoorpAcademy/baucis)
[![codecov](https://codecov.io/gh/CoorpAcademy/baucis/branch/master/graph/badge.svg)](https://codecov.io/gh/CoorpAcademy/baucis)

> Build scalable REST APIs using the open source tools and standards you and your team already know — *Mongoose, Express, and Node.js streams*.

Baucis takes the boilerplate out of building and maintaining scalable [HATEOAS](https://en.wikipedia.org/wiki/HATEOAS)/[Level 3](http://martinfowler.com/articles/richardsonMaturityModel.html) REST APIs.


:warning: This is a **fork** from the *Coorpacademy Tech Team* :warning:

This is so far intended for intern use

## About this fork :book:

This fork adapt the internal structure of baucis, and the plugin system.
It consist in a monorepros hosting the main baucis plugins (adapted to the new system).

One other major changes is that it enable you to inject the version of express or mongoose you want,
getting ride of the mongoose <5 constraint.

Packages are release for the moment under `@coorpacademy/` scope, but this is subject to change if the fork
happens to be working and interest span out coorpacademy

## Features :hammer_and_wrench:

 * Automatically build controllers from your Mongoose schemas, then easily configure them.
 * Built on Express 4 so adding custom middleware is a snap.  Compatible with existing Express middleware.
 * Fully takes advantage of Node.js streaming to nimbly process large datasets.
 * Implements the HTTP specification according to the specifications ([RFC 7231](http://tools.ietf.org/rfcmarkup/7231), etc.)
 * Widely compatible with a variety of front end frameworks.
 * Perform rich queries of the API using an expressive JSON syntax via query string.
 * Supports geolocation and full text search.
 * Version your API using semver.
 * Automatically generate interactive Swagger documentation for the API.
 * Highly customizable, simple interface.  Can be extended with plugins.
 * Compatible with MongoDB 2.x and 3.x
 * Over 140 Mocha.js tests in addition to Express' and Mongoose's.


## Getting Started :rocket:

To install:

    npm install --save @coorpacademy/baucis

It's easy to create an API from a Mongoose model:

``` javascript
const mongoose = require('mongoose');
const express = require('express');
const baucis = require('@coorpacademy/baucis')(mongoose, express)
// Create a mongoose schema.
const Vegetable = new mongoose.Schema({ name: String });
// Register new models with mongoose.
mongoose.model('vegetable', Vegetable);
// Create a simple controller.  By default these HTTP methods
// are activated: HEAD, GET, POST, PUT, DELETE
baucis.rest('vegetable');
// Create the app and listen for API requests
const app = express();
app.use('/api', baucis());
app.listen(4312);
```

That's it!  Now you have an RFC-compliant, flexible, and semantically rich API dealing with vegetables.  You could access it with URLs like `http://localhost:4312/api/vegetables`.  CRUD is supported using [GET, PUT, POST, and DELETE](https://github.com/wprl/baucis/wiki/HTTP-Verbs).


## Examples :national_park:

Some new example is present in the [examples](./examples) folder. It make use of swagger plugins so you can directly
have some explorer disponible when you run it.

## Documentation :closed_book:

[Check out the Doc](./documention) for more in-depth information about baucis.

Check the [change log](CHANGES.md) for info on recently implemented features.


CHANGES.md
# Baucis Change Log


***Coorpacademy Fork took place at v1.6.8***
--------

Changes from v1.0 to v1.6.8 were not documented

## v1.0.0-candidate.11

Baucis was awarded the MongoDB Open Source Hack Contest 2014!

Update readme and roadmap.

## v1.0.0-candidate.10

Fix some typos.


## v1.0.0-candidate.9

Add more info about `baucis.rest` to the README.


## v1.0.0-candidate.8

Use caret for dependencies.  If you're using an older npm, let's talk about a fork or other migration path.


## v1.0.0-candidate.7

README tweaks.  More tests.


## v1.0.0-candidate.6

Minor tweaks.


## v1.0.0-candidate.5

Revisit all error conditions.  Make 422 error response body an array.


## v1.0.0-candidate.4

Fix issue #183.


## v1.0.0-candidate.3

Update baucis-error version.


## v1.0.0-candidate.2

Make 200 default for empty collection/search.  Split out baucis-errors and baucis-vivify.


## v1.0.0-candidate.1

Fix #149.  Fix #181.  Move handleErrors from Api to Controller.  Add parentController method.  Add emptyCollection method to controller.  It is set to 204 by defult, but can be set to 200, 204, or 404 to set how collection endpoint requests handle no matching documents per controller.


## v1.0.0-candidate.0

Use published version of baucis-json.


## v1.0.0-prerelease.9

Allow turning off error handling per API.  Fix #178.

    baucis().handleErrors(false);


## v1.0.0-prerelease.8

Correctly handle errors in user streams.  Improve error handling.  Fix #177.


## v1.0.0-prerelease.7

Rename `baucisPath` to `fragment`.  Use a decorator for Mongoose models, rather than a wrapper object.  Wrap `mongoose.model` to add decorator to all models.  Move `lastModified` and `locking` to model objects.


## v1.0.0-prerelease.6

Move parsers/formatters from Api to baucis object.


## v1.0.0-prerelease.5

Add PATENTS file.  A little more cleaning up.


## v1.0.0-prerelease.4

Remove `request.baucis.controller`.  Remove `controller.api`.


## v1.0.0-prerelease.3

Express 4!
del -> delete
Controller activation is no longer required (again)
API activation is not required
Controller properties can change at runtime, and the API will now reflect those changes
Remove X-Powered-By for all baucis routes


## v1.0.0-prerelease.2

Merge to master and make small update to the README.


## v1.0.0-prerelease.1

Add Model.  Change X-Baucis-Update-Operator to Update-Operator.  Begin refactoring Api into smaller decorators.


## v0.20.5

Update to the latest baucis-json.  Fixes an issue where POSTing or PUTing some kinds of malformed JSON bodies could crash the instance.  Now sends a 400 Bad Request for malformed JSON.  Fix #168


## v0.20.4

Check additional opportunities to send 400 Bad Request.


## v0.20.3

README improvements.


## v0.20.2

Fix bug that overwrote Express' parent property.  Please use the `controller.children` property instead.  (Fixes #161.)


## v0.20.1

Fix publish and some typos.


## v0.20.0

Add `controller.vivify` method for embedding subcontrollers.  Deprecate `basePath` controller option.  Update configuration to "methods over values."  Move baucis errors from `baucis.errors` to `baucis.Error`.  Enable relations by default.  Improve & update README.


## v0.19.2

Handle duplicate key error as a 422.  (Fix #160).


## v0.19.1

Fix #159.


## v0.19.0

Add context to incoming and outgoing user streams. (Fix #151).


## v0.18.4

Internal refactor to use protected pipeline method for internal as well as external stream interfaces.


## v0.18.3

Update to use the latest `baucis-json`.


## v0.18.2

Don't try to set the Last-Modified header if the feature isn't enabled (fix #157).


## v0.18.1

Fix an issue where the request's content type would not be recognized when it used a MIME type parameter. (#152)


## v0.18.0

Use the mongoose `toCollectionName` util method, instead of the lingo module.


## v0.17.0

Deprecate `documents` stage of middleware.


## v0.16.21

Update dependency for deco.


## v0.16.20

Updated outdated dependencies.


## v0.16.19

Improve error messages, correct some status codes, fix issue #147.


## v0.16.18

Discriminators - correctly handle POST.  The correct model is used based on the `__t` that is passed in the request body.


## v0.16.17

Use tilde instead of caret so older npms don't fall over!


## v0.16.16

Move JSON parser and formatter to a separate module.


## v0.16.15

Fix outdated require.


## v0.16.14

Add distinct request query option.


## v0.16.13

Fix missing dependency on event-stream.


## v0.16.12

Send 415 on unparseable request content type.  Also added experimental `Api.setParser` method.


## v0.16.11

Ensure 406 is sent on requests with Accept header that cannot be fulfilled.  Fix bug with calling `api.setFormatter`.


## v0.16.10

Allow passing in map function to incoming/outgoing user streams.  It's turned into a stream internally.


## v0.16.9

Add experimental `Api.setFormatter` method.


## v0.16.8

Fix incorrect Accept header.


## v0.16.7

Allow passing in a mongoose model to the `model` controller option, in addition to string name.


## v0.16.6

Update dependency for Express to 3.5.


## v0.16.5

Fix issue #142.  Also, use event-stream over highland.


## v0.16.4

Fix issue #141


## v0.16.3

Fix issue #131


## v0.16.2

Add documentation for streaming and clean up documentation.


## v0.16.1

Republish failed `npm publish`.


## v0.16.0

Streamy goodness!  Documentation will be updated for v0.16.2.  Things are in flux...


## v0.16.0-alpha

Stream documents to manage resources more efficiently.  (Performance tests forthcoming.)

The documentation has not been updated yet... public interface subject to change.  Suggestions welcome!


## v0.15.1

Send a descriptive body for validation errors.


## v0.15.0

Send 422 on validation error.


## v0.14.2

Remove stale calls to `connect.utils.merge`.


## v0.14.1

Update to use latest `deco` features.  Other additional refactoring.


## v0.14.0


src/controller/send.js
// __Dependencies__
const crypto = require('crypto');
const eventStream = require('event-stream');
const RestError = require('rest-error');

// __Module Definition__
module.exports = function(options, protect) {
  const baucis = require('../..');
  const controller = this;
  const lastModifiedPath = controller.model().lastModified();
  const trailers = {};

  // __Private Module Members__
  // Format the Trailer header.
  function addTrailer(response, header) {
    const current = response.get('Trailer');
    if (!current) response.set('Trailer', header);
    else response.set('Trailer', `${current}, ${header}`);
  }
  // A map that is used to create empty response body.
  function empty(context, callback) {
    callback(null, '');
  }
  // Map contexts back into documents.
  function redoc(context, callback) {
    callback(null, context.doc);
  }
  // Generate a respone Etag from a context.
  function etag(response, useTrailer) {
    if (useTrailer) {
      addTrailer(response, 'Etag');
      response.set('Transfer-Encoding', 'chunked');
    }

    const hash = crypto.createHash('md5');

    return eventStream.through(
      function(chunk) {
        hash.update(chunk);
        this.emit('data', chunk);
      },
      function() {
        if (useTrailer) {
          trailers.Etag = `"${hash.digest('hex')}"`;
        } else {
          response.set('Etag', `"${hash.digest('hex')}"`);
        }

        this.emit('end');
      }
    );
  }

  function etagImmediate(response) {
    const hash = crypto.createHash('md5');

    return eventStream.through(
      function(chunk) {
        hash.update(JSON.stringify(chunk));
        response.set('Etag', `"${hash.digest('hex')}"`);
        this.emit('data', chunk);
      },
      function() {
        this.emit('end');
      }
    );
  }
  // Generate a Last-Modified header/trailer
  function lastModified(response, useTrailer) {
    if (useTrailer) {
      addTrailer(response, 'Last-Modified');
      response.set('Transfer-Encoding', 'chunked');
    }

    let latest = null;

    return eventStream.through(
      function(context) {
        if (!context) return;
        if (!context.doc) return this.emit('data', context);
        if (!context.doc.get) return this.emit('data', context);

        const current = context.doc.get(lastModifiedPath);
        if (latest === null) latest = current;
        else latest = new Date(Math.max(latest, current));
        if (!useTrailer) {
          response.set('Last-Modified', latest.toUTCString());
        }
        this.emit('data', context);
      },
      function() {
        if (useTrailer) {
          if (latest) trailers['Last-Modified'] = latest.toUTCString();
        }

        this.emit('end');
      }
    );
  }

  // Build a reduce stream.
  function reduce(accumulated, f) {
    return eventStream.through(
      function(context) {
        accumulated = f(accumulated, context);
      },
      function() {
        this.emit('data', accumulated);
        this.emit('end');
      }
    );
  }
  // Count emissions.
  function count() {
    return reduce(0, function(a, b) {
      return a + 1;
    });
  }

  // If counting get the count and send it back directly.
  protect.finalize(function(request, response, next) {
    if (!request.baucis.count) return next();

    request.baucis.query.count(function(error, n) {
      if (error) return next(error);
      response.removeHeader('Transfer-Encoding');
      return response.json(n); // TODO support other content types
    });
  });

  // If not counting, create the basic stream pipeline.
  protect.finalize('collection', 'all', function(request, response, next) {
    let count = 0;
    const documents = request.baucis.documents;
    const pipeline = protect.pipeline(next);
    request.baucis.send = pipeline;
    // If documents were set in the baucis hash, use them.
    if (documents) pipeline(eventStream.readArray([].concat(documents)));
    else {
      // Otherwise, stream the relevant documents from Mongo, based on constructed query.
      pipeline(request.baucis.query.cursor());
    }

    // Map documents to contexts.
    pipeline(function(doc, callback) {
      callback(null, {doc, incoming: null});
    });
    // Check for not found.
    pipeline(
      eventStream.through(
        function(context) {
          count += 1;
          this.emit('data', context);
        },
        function() {
          if (count > 0) return this.emit('end');

          const status = controller.emptyCollection();
          response.status(status);

          if (status === 204) {
            response.removeHeader('Trailer');
            return this.emit('end');
          }
          if (status === 200) {
            response.removeHeader('Transfer-Encoding');
            response.removeHeader('Trailer');
            response.json([]); // TODO other content types
            this.emit('end');
            return;
          }

          this.emit('error', RestError.NotFound());
        }
      )
    );
    // Apply user streams.
    pipeline(request.baucis.outgoing());

    // Set the document formatter based on the Accept header of the request.
    baucis.formatters(response, function(error, formatter) {
      if (error) return next(error);
      request.baucis.formatter = formatter;
      next();
    });
  });

  protect.finalize('instance', 'all', function(request, response, next) {
    let count = 0;
    const documents = request.baucis.documents;
    const pipeline = protect.pipeline(next);
    request.baucis.send = pipeline;
    // If documents were set in the baucis hash, use them.
    if (documents) {
      pipeline(eventStream.readArray([].concat(documents)));
    } else {
      // Otherwise, stream the relevant documents from Mongo, based on constructed query.
      pipeline(request.baucis.query.cursor());
    }

    // Map documents to contexts.
    pipeline(function(doc, callback) {
      callback(null, {doc, incoming: null});
    });
    // Check for not found.
    pipeline(
      eventStream.through(
        function(context) {
          count += 1;
          this.emit('data', context);
        },
        function() {
          if (count > 0) return this.emit('end');
          this.emit('error', RestError.NotFound());
        }
      )
    );
    // Apply user streams.
    pipeline(request.baucis.outgoing());

    // Set the document formatter based on the Accept header of the request.
    baucis.formatters(response, function(error, formatter) {
      if (error) return next(error);
      request.baucis.formatter = formatter;
      next();
    });
  });

  // OPTIONS // TODO Express' extra handling for OPTIONS conflicts with baucis
  // TODO maybe send method names in body
  // controller.options(function (request, response, next) {
  //   console.log('here')
  //   request.baucis.send(empty);
  //   next();
  // });

  // HEAD
  protect.finalize('instance', 'head', function(request, response, next) {
    if (lastModifiedPath) {
      request.baucis.send(lastModified(response, false));
    }

    request.baucis.send(redoc);
    request.baucis.send(etagImmediate(response));
    request.baucis.send(request.baucis.formatter());
    request.baucis.send(empty);
    next();
  });

  protect.finalize('collection', 'head', function(request, response, next) {
    if (lastModifiedPath) {
      request.baucis.send(lastModified(response, false));
    }

    request.baucis.send(redoc);
    request.baucis.send(request.baucis.formatter(true));
    request.baucis.send(etag(response, false));
    request.baucis.send(empty);
    next();
  });

  // GET
  protect.finalize('instance', 'get', function(request, response, next) {
    if (lastModifiedPath) {
      request.baucis.send(lastModified(response, false));
    }

    request.baucis.send(redoc);
    request.baucis.send(etagImmediate(response));
    request.baucis.send(request.baucis.formatter());
    next();
  });

  protect.finalize('collection', 'get', function(request, response, next) {
    if (lastModifiedPath) {
      request.baucis.send(lastModified(response, true));
    }

    if (request.baucis.count) {
      request.baucis.send(count());
      request.baucis.send(eventStream.stringify());
    } else {
      request.baucis.send(redoc);
      request.baucis.send(request.baucis.formatter(true));
    }

    request.baucis.send(etag(response, true));
    next();
  });

  // POST
  protect.finalize('collection', 'post', function(request, response, next) {
    request.baucis.send(redoc);
    request.baucis.send(request.baucis.formatter());
    next();
  });

  // PUT
  protect.finalize('put', function(request, response, next) {
    request.baucis.send(redoc);
    request.baucis.send(request.baucis.formatter());
    next();
  });

  // DELETE
  protect.finalize('delete', function(request, response, next) {
    // Remove each document from the database.
    request.baucis.send(function(context, callback) {
      context.doc.remove(callback);
    });
    // Respond with the count of deleted documents.
    request.baucis.send(count());
    request.baucis.send(eventStream.stringify());
    next();
  });

  protect.finalize(function(request, response, next) {
    request.baucis.send().pipe(
      eventStream.through(
        function(chunk) {
          response.write(chunk);
        },
        function() {
          response.addTrailers(trailers);
          response.end();
          this.emit('end');
        }
      )
    );
  });
};
